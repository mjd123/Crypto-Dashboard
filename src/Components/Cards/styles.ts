import styled from "styled-components/macro";
import { PageText as CardText } from "../../helpers/PageText";
import { Wrapper as CardWrapper } from "../../helpers/Wrapper";
import { device } from "../../styles/Breakpoints";

export const Text = styled(CardText)`
  color: ${(props) => (props.color ? props.color : "#ffffff")};
  font-size: ${(props) => (props.fontSize ? props.fontSize : ".9em")};
  white-space: nowrap;
`;

export const Wrapper = styled(CardWrapper)`
  //display: flex;
  position: relative;
  justify-content: ${(props) => props.justifyContent};
  align-items: ${(props) => props.justifyContent};
  padding: 10px;
  grid-row: 2 / 4;
  grid-column: 2;
  box-shadow: inset 0 0 0 100vmax rgba(0, 0, 0, 0.6);
  border-radius: 25px;
  margin: 20px;
  justify-self: end;
  width: fit-content;
  height: fit-content;
  min-height: 300px;
  min-width: 300px;
  .ant-card :not(.expanded) {
    display: grid;
    grid-template: 1fr /35px auto;
    align-items: center;
    justify-content: space-evenly;
    text-align: center;
    padding: 10px;

    .ant-card-body {
      padding: 0;
      display: grid;
      grid-template: 1fr /0 1fr 1fr 1fr;
      align-items: center;
      grid-column: 2;
      grid-row: 1;
      width: 100%;
    }
  }

  ${device.md`
    grid-column: 1 / span 2;
    margin:auto;
    justify-self:center;
    height:fit-content;
  `};
`;

export const CardTextWrapper = styled(CardWrapper)`
  display: ${(props) => props.display};
  flex-direction: ${(props) => props.flexDirection};
  justify-content: ${(props) => props.justifyContent};
  height: ${(props) => (props.height ? props.height : "100%")};
  width: 100%;
  grid-row: ${(props) => props.gridRow};
  margin-bottom: ${(props) => props.marginBottom};
  margin-left: ${(props) => props.marginLeft};
  margin-right: ${(props) => props.marginRight};
`;

export const Image = styled.img`
  max-width: 15px;
`;
