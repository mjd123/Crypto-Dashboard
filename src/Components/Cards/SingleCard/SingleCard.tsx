import { Text, CardTextWrapper } from "../styles";
import "antd/dist/antd.css";
import { Card } from "antd";
import Chart from "./Chart/Chart";
import { memo, useEffect, useState } from "react";
import CloseButton from "./CloseButton/CloseButton";
import DateButtons from "./DateButtons/DateButtons";
import PriceStats from "./PriceStats/PriceStats";

export interface IndividualCardProps {
  coinImage?: string | undefined;
  text: string;
  symbol: string;
  price: string | number;
  priceChange: number;
  history?: any;
  expanded?: boolean | undefined;
  totalVolume?: number;
  dayRange?: { low: number; high: number };
  keyId: number;
}

export const timeframeKeys: string[] = [
  "yesterday",
  "sevenDays",
  "oneMonth",
  "oneYearToDate",
  "oneYear",
  "fiveYears",
  "All",
];

const IndividualCard = memo(
  ({
    coinImage,
    text,
    symbol,
    price,
    priceChange,
    history,
    expanded,
    totalVolume,
    dayRange,
    keyId,
  }: IndividualCardProps) => {
    const [priceData, setPriceData] = useState<
      [{ price: number; priceDate: string | undefined }] | []
    >([]);
    const [expandCard, setExpandCard] = useState<boolean | undefined>(false);
    const [timeframe, setTimeframe] = useState("yesterday");
    const [defaultButtonHighlight, setDefaultButtonHighlight] = useState(true);

    useEffect(() => {
      //reset price data on props change
      setPriceData([]);
    }, [history, timeframe]);

    useEffect(() => {
      // parsing props.history date and price into human readable version
      history &&
        history[timeframe]?.map((x: number[], i: number) => {
          //setPriceData("");
          return setPriceData((priceData): any => [
            ...priceData,
            {
              priceDate: x[0], //new Date(x[0]).toISOString().slice(5, 10),
              price: x[1], // need to keep decimal places
            },
          ]);
        });
    }, [history, timeframe]);

    useEffect(() => {
      if (expanded) {
        setExpandCard(expanded);
      } else {
        setExpandCard(false);
      }
    }, [expanded]);

    // recieves data from DateButton to refresh props sent to Chart
    const timeframeChange = (data: string) => {
      setPriceData([]);
      setTimeframe(data);
      setDefaultButtonHighlight(false);
    };

    const gridStyleExpanded: {} = {
      display: "grid",
      gridTemplate: "50px  1fr/ 1fr",
      justifyItems: "center",
      alignItems: "center",
      textAlign: "center",
      padding: "10px",
      width: "100%",
      height: "100%",
      borderRadius: "10px",
    };

    const expandImageStyle = {
      gridArea: 0,
      gridRow: 0,
      alignSelf: "center",
      justifySelf: "center",
      maxWidth: "30px",
    };

    const cardBodyExpanded = {
      padding: 0,
      height: "100%",
      width: "100%",
      display: "grid",
      gridTemplateRows: "0px auto",
      gridTemplateColumns: "100%",
    };

    return (
      <>
        <Card
          data-testid="card"
          className={expandCard ? "expanded" : ""}
          style={
            expandCard
              ? gridStyleExpanded
              : {
                  borderRadius:
                    keyId === 0
                      ? "10px 10px 0 0"
                      : keyId === 7
                      ? "0 0 10px 10px"
                      : "",
                }
          }
          bodyStyle={expandCard ? cardBodyExpanded : {}}
          hoverable={expandCard ? false : true}
          cover={
            <img
              style={expandCard ? expandImageStyle : { width: "30px" }}
              alt="coin"
              src={coinImage}
            />
          }
        >
          {expandCard && <CloseButton />}
          <CardTextWrapper
            display="flex"
            flexDirection="column"
            marginBottom="0"
          >
            {!expandCard && <Text color="black">{text} </Text>}
            <Text color="black">{symbol.toUpperCase()}</Text>
          </CardTextWrapper>
          <CardTextWrapper
            className="chart-wrapper"
            marginBottom={expandCard ? "1.5em" : "0"}
            height={"auto"}
          >
            {history && (
              <Chart
                expanded={expandCard}
                history={priceData}
                priceChange={priceChange}
                timeframe={timeframe}
              />
            )}
          </CardTextWrapper>

          <CardTextWrapper
            className="price-wrapper"
            display="flex"
            flexDirection="column"
            gridRow={expandCard ? 3 : 0}
            marginBottom={expandCard ? "1.5em" : "0"}
          >
            <Text color="black">${price.toLocaleString("en")}</Text>
            <Text color={priceChange >= 0 ? "green" : "red"}>
              {priceChange >= 0 && "+"}
              {priceChange.toFixed(2)}%
            </Text>
          </CardTextWrapper>
          {expandCard && totalVolume && history && (
            <>
              <CardTextWrapper
                className="date-button-wrapper"
                display="flex"
                flexDirection="row"
                justifyContent="space-evenly"
                marginBottom="1.5em"
              >
                {timeframeKeys.map((item: string, i: number) => {
                  return (
                    <DateButtons
                      key={i}
                      timeframe={item}
                      timeframeData={timeframeChange}
                      initial={
                        item === "yesterday" && defaultButtonHighlight
                          ? true
                          : false
                      }
                      disabled={item === timeframe ? true : false}
                    />
                  );
                })}
              </CardTextWrapper>
              <PriceStats
                yesterdayPriceHistory={history.yesterday}
                volume={totalVolume}
                range={{
                  low: dayRange?.low,
                  high: dayRange?.high,
                }}
              />
            </>
          )}
        </Card>
      </>
    );
  }
);

export default IndividualCard;
