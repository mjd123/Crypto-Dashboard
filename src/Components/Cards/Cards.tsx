import { useEffect, useState, useContext, useRef } from "react";
import SingleCard from "./SingleCard/SingleCard";
import { Wrapper } from "./styles";
import { Row } from "antd";
import "antd/dist/antd.css";
import { CardContext } from "../../contexts/ToggleCardExpandContext";
import { AnimatePresence, motion } from "framer-motion";
import Loading from "../../helpers/Loading";
import { gql, useQuery } from "@apollo/client";

const COIN_DATA = gql`
  query {
    coinData {
      id
      symbol
      name
      image
      current_price
      high_24h
      low_24h
      price_change_24h
      price_change_percentage_24h
      ath
      total_volume
    }
  }
`;

const Cards = () => {
  const { cardStatus, toggleStatus } = useContext(CardContext);
  const [rowClicked, setRowClicked] = useState<number | null>(null);
  const [fadeComponent, setFadeComponent] = useState(false);
  const ref = useRef<HTMLInputElement | null>(null);
  const { loading, error, data } = useQuery(COIN_DATA, { pollInterval: 9000 });

  const expandRow = (
    e: MouseEvent | TouchEvent | PointerEvent,
    index: number
  ) => {
    e.preventDefault();
    e.stopPropagation();
    toggleStatus(index);
    setRowClicked(index);
  };

  return (
    <motion.div
      ref={ref}
      animate={{ width: " 100% " }}
      transition={{
        duration: 0.3,
        ease: "easeInOut",
      }}
    >
      <Wrapper className="cards-wrapper">
        <Loading loading={loading} />
        {!loading &&
          data &&
          error === undefined &&
          data.coinData.map((coin: Record<string, any>, i: number) => {
            return (
              <div key={i}>
                <AnimatePresence onExitComplete={() => setFadeComponent(true)}>
                  {cardStatus === null && !fadeComponent && (
                    <motion.div
                      onTap={(e) => expandRow(e, i)}
                      initial={{ opacity: 0 }}
                      animate={{ opacity: 1, scale: 1 }}
                      exit={{ opacity: 0 }}
                      transition={{
                        duration: 0.5,
                        ease: "easeInOut",
                      }}
                    >
                      <Row
                        gutter={0}
                        style={
                          i === 0
                            ? { borderRadius: "10px 10px 0 0 ", width: "100%" }
                            : { width: "100%" }
                        }
                        className="single-card-wrapper"
                      >
                        <SingleCard
                          id={coin.id}
                          keyId={i}
                          coinImage={coin.image}
                          price={coin.current_price}
                          priceChange={coin.price_change_percentage_24h}
                          totalVolume={coin.total_volume}
                          text={coin.name}
                          symbol={coin.symbol}
                        />
                      </Row>
                    </motion.div>
                  )}
                </AnimatePresence>

                <AnimatePresence onExitComplete={() => setFadeComponent(false)}>
                  {rowClicked === i && cardStatus === i && fadeComponent && (
                    <motion.div
                      onTap={(e) => expandRow(e, i)}
                      initial={{ opacity: 0 }}
                      animate={{ opacity: 1 }}
                      exit={{ opacity: 0 }}
                      transition={{
                        ease: "easeInOut",
                        duration: 0.3,
                        staggerChildren: 3,
                      }}
                    >
                      <Row
                        gutter={0}
                        style={{ height: "100%" }}
                        className="single-card-wrapper"
                      >
                        <SingleCard
                          id={coin.id}
                          expanded={rowClicked === i ? true : false}
                          keyId={i}
                          coinImage={coin.image}
                          price={coin.current_price}
                          priceChange={coin.price_change_percentage_24h}
                          text={coin.name}
                          symbol={coin.symbol}
                          totalVolume={coin.total_volume}
                          dayRange={{
                            low: coin.low_24h || 0,
                            high: coin.high_24h || 0,
                          }}
                        />
                      </Row>
                    </motion.div>
                  )}
                </AnimatePresence>
              </div>
            );
          })}
      </Wrapper>
    </motion.div>
  );
};

export default Cards;
