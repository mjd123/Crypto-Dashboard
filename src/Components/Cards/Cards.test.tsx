import React from "react";
import { shallow, mount } from "enzyme";
import Cards from "./Cards";
import useFetch from "../../hooks/useFetchData";

import toJson from "enzyme-to-json";
import SingleCard from "./SingleCard/SingleCard";

const mockData = [
  {
    id: "bitcoin",
    symbol: "btc",
    name: "Bitcoin",
    image:
      "https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579",
    current_price: 61173,
    market_cap: 1153049002571,
    market_cap_rank: 1,
    fully_diluted_valuation: 1284097717417,
    total_volume: 43757353911,
    high_24h: 61406,
    low_24h: 58369,
    price_change_24h: 2515.05,
    price_change_percentage_24h: 4.28768,
    market_cap_change_24h: 46031325301,
    market_cap_change_percentage_24h: 4.15814,
    circulating_supply: 18856843,
    total_supply: 21000000,
    max_supply: 21000000,
    ath: 67277,
    ath_change_percentage: -9.11356,
    ath_date: "2021-10-20T14:54:17.702Z",
    atl: 67.81,
    atl_change_percentage: 90073.1045,
    atl_date: "2013-07-06T00:00:00.000Z",
    roi: null,
    last_updated: "2021-10-28T12:11:07.552Z",
  },
];

jest.mock("../../hooks/useFetchData", () => ({
  __esModule: true,
  default: () => mockData,
}));

describe("Card wont fail without component", () => {
  const component = mount(<Cards />);
  test("Cards component renders without crashing", () => {
    const children = component.find("Wrapper");

    expect(children).toHaveLength(1);
  });
});
