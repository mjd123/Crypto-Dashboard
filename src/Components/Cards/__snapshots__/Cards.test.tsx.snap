// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card wont fail without component Cards component renders without crashing 1`] = `ReactWrapper {}`;
